test_command_fail_1 fails on debian testing py 3.9 (only)

```
_____________________________ test_command_fail_1 ______________________________

    def test_command_fail_1():
        # Expect that the return code of a failing command is caught,
        # that None is returned as result.
        bc = BatchedCommand(
            cmd=py2cmd(
                """
    print("something")
    exit(3)
                """))
    
        # Send something to start the process
        result = bc("line one")
        assert_equal(bc.return_code, None)
        assert_equal(result, "something")
        result = bc("line two")
>       assert_equal(bc.return_code, 3)

datalad/tests/test_cmd.py:165: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

first = None, second = 3, msg = None

    def assert_equal(first, second, msg=None):
        if msg is None:
>           assert first == second
E           assert None == 3
E             +None
E             -3

datalad/tests/utils_pytest.py:107: AssertionError
```

as was added in 9f066efa6a - @christian-monch any ideas ?

--- a/changelog.d/pr-7068.md
+++ b/changelog.d/pr-7068.md
@@ -1,6 +1,3 @@
 ### Internal
 
-- Ensure that `BatchedCommandError` is raised if the subprocesses of
-  `BatchedCommand` fails or raises a `CommandError`.  [PR
-  #7068](https://github.com/datalad/datalad/pull/7068) (by
-  [@christian-monch](https://github.com/christian-monch))
+- Ensure that `BatchedCommandError` is raised if the subprocesses of `BatchedCommand` fails or raises a `CommandError`.  [PR #7068](https://github.com/datalad/datalad/pull/7068) (by [@christian-monch](https://github.com/christian-monch))

--- a/changelog.d/pr-7068.md
+++ b/changelog.d/pr-7068.md
@@ -1,0 +1,6 @@
+### Internal
+
+- Ensure that `BatchedCommandError` is raised if the subprocesses of
+  `BatchedCommand` fails or raises a `CommandError`.  [PR
+  #7068](https://github.com/datalad/datalad/pull/7068) (by
+  [@christian-monch](https://github.com/christian-monch))

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -401,7 +401,7 @@ class BatchedCommand(SafeDelCloseMixin):
             if not self.process_running():
                 self._initialize()
 
-            # Remember request end send it to subprocess
+            # Remember request and send it to subprocess
             if not isinstance(request, str):
                 request = ' '.join(request)
             self.stdin_queue.put((request + "\n").encode())

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -68,6 +68,30 @@ from datalad.utils import (
 )
 
 
+class BatchedCommandError(CommandError):
+    def __init__(self,
+                 cmd="",
+                 request="",
+                 msg="",
+                 code=None,
+                 stdout="",
+                 stderr="",
+                 cwd=None,
+                 **kwargs):
+
+        CommandError.__init__(
+            self,
+            cmd=cmd,
+            msg=msg,
+            code=code,
+            stdout=stdout,
+            stderr=stderr,
+            cwd=cwd,
+            **kwargs
+        )
+        self.request = request
+
+
 lgr = logging.getLogger('datalad.cmd')
 
 # TODO unused?
@@ -200,6 +224,7 @@ class BatchedCommand(SafeDelCloseMixin):
         self.wait_timed_out = None
         self.return_code = None
         self._abandon_cache = None
+        self.last_request: Optional[str] = None
 
         self._active = 0
         self._active_last = _now()
@@ -259,6 +284,7 @@ class BatchedCommand(SafeDelCloseMixin):
         self.stderr_output = b""
         self.wait_timed_out = None
         self.return_code = None
+        self.last_request = None
 
         self.runner = WitlessRunner(
             cwd=self.path,
@@ -282,7 +308,19 @@ class BatchedCommand(SafeDelCloseMixin):
 
     def process_running(self) -> bool:
         if self.runner:
-            return self.generator.runner.process.poll() is None
+            result = self.generator.runner.process.poll()
+            if result is None:
+                return True
+            self.return_code = result
+            if result != 0:
+                self.runner = None
+                raise BatchedCommandError(
+                    cmd=" ".join(self.command),
+                    request=self.last_request,
+                    msg=f"{type(self).__name__}: exited with {result} after "
+                        f"request: {self.last_request}",
+                    code=result
+                )
         return False
 
     def __call__(self,
@@ -334,11 +372,6 @@ class BatchedCommand(SafeDelCloseMixin):
                         self.return_code = self.generator.return_code
                         self.runner = None
 
-        except CommandError as command_error:
-            # The command exited with a non-zero return code
-            lgr.error("%s: command error: %s", self, command_error)
-            self.return_code = command_error.code
-            self.runner = None
 
         finally:
             self._active -= 1
@@ -353,9 +386,10 @@ class BatchedCommand(SafeDelCloseMixin):
             if not self.process_running():
                 self._initialize()
 
-            # Send request to subprocess
+            # Remember request end send it to subprocess
             if not isinstance(request, str):
                 request = ' '.join(request)
+            self.last_request = request
             self.stdin_queue.put((request + "\n").encode())
 
             # Get the response from the generator. We only consider

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -336,7 +336,7 @@ class BatchedCommand(SafeDelCloseMixin):
                     msg=f"{type(self).__name__}: exited with {result} after "
                         f"request: {self.last_request}",
                     code=result
-                )
+                ) from CommandError
         return False
 
     def __call__(self,
@@ -402,7 +402,7 @@ class BatchedCommand(SafeDelCloseMixin):
                 stderr=command_error.stderr,
                 cwd=command_error.cwd,
                 **command_error.kwargs
-            )
+            ) from command_error
 
         finally:
             self._active -= 1

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -222,7 +222,7 @@ class BatchedCommand(SafeDelCloseMixin):
         self.runner: Optional[WitlessRunner] = None
         self.encoding = None
         self.wait_timed_out = None
-        self.return_code = None
+        self.return_code: Optional[int] = None
         self._abandon_cache = None
         self.last_request: Optional[str] = None
 

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -68,17 +68,33 @@ from datalad.utils import (
 )
 
 
+__docformat__ = "restructuredtext"
+
+
 class BatchedCommandError(CommandError):
     def __init__(self,
                  cmd="",
-                 request="",
+                 last_processed_request="",
                  msg="",
                  code=None,
                  stdout="",
                  stderr="",
                  cwd=None,
                  **kwargs):
+        """
+        This exception extends a CommandError that is raised by the command,
+        that is executed by `BatchedCommand`. It extends the `CommandError` by
+        `last_processed_request`. This attribute contains the last request, i.e.
+        argument to `BatchedCommand.__call__()`, that was successfully
+        processed, i.e. for which a result was received from the command (that
+        does not imply that the result was positive).
 
+        :param last_processed_request: the last request for which a response was
+            received from the underlying command. This could be used to restart
+            an interrupted process.
+
+        For all other arguments see `CommandError`.
+        """
         CommandError.__init__(
             self,
             cmd=cmd,
@@ -89,7 +105,7 @@ class BatchedCommandError(CommandError):
             cwd=cwd,
             **kwargs
         )
-        self.request = request
+        self.last_processed_request = last_processed_request
 
 
 lgr = logging.getLogger('datalad.cmd')
@@ -316,7 +332,7 @@ class BatchedCommand(SafeDelCloseMixin):
             if result != 0:
                 raise BatchedCommandError(
                     cmd=" ".join(self.command),
-                    request=self.last_request,
+                    last_processed_request=self.last_request,
                     msg=f"{type(self).__name__}: exited with {result} after "
                         f"request: {self.last_request}",
                     code=result
@@ -379,7 +395,7 @@ class BatchedCommand(SafeDelCloseMixin):
             self.return_code = command_error.code
             raise BatchedCommandError(
                 cmd=command_error.cmd,
-                request=self.last_request,
+                last_processed_request=self.last_request,
                 msg=command_error.msg,
                 code=command_error.code,
                 stdout=command_error.stdout,

--- a/datalad/cmd.py
+++ b/datalad/cmd.py
@@ -312,8 +312,8 @@ class BatchedCommand(SafeDelCloseMixin):
             if result is None:
                 return True
             self.return_code = result
+            self.runner = None
             if result != 0:
-                self.runner = None
                 raise BatchedCommandError(
                     cmd=" ".join(self.command),
                     request=self.last_request,
@@ -364,6 +364,7 @@ class BatchedCommand(SafeDelCloseMixin):
                 while True:
                     try:
                         responses.append(self.process_request(request))
+                        self.last_request = request
                         break
                     except StopIteration:
                         # The process finished executing, store the last return
@@ -372,6 +373,20 @@ class BatchedCommand(SafeDelCloseMixin):
                         self.return_code = self.generator.return_code
                         self.runner = None
 
+        except CommandError as command_error:
+            # Convert CommandError into BatchedCommandError
+            self.runner = None
+            self.return_code = command_error.code
+            raise BatchedCommandError(
+                cmd=command_error.cmd,
+                request=self.last_request,
+                msg=command_error.msg,
+                code=command_error.code,
+                stdout=command_error.stdout,
+                stderr=command_error.stderr,
+                cwd=command_error.cwd,
+                **command_error.kwargs
+            )
 
         finally:
             self._active -= 1
@@ -389,7 +404,6 @@ class BatchedCommand(SafeDelCloseMixin):
             # Remember request end send it to subprocess
             if not isinstance(request, str):
                 request = ' '.join(request)
-            self.last_request = request
             self.stdin_queue.put((request + "\n").encode())
 
             # Get the response from the generator. We only consider

--- a/datalad/support/annexrepo.py
+++ b/datalad/support/annexrepo.py
@@ -2766,8 +2766,9 @@ class AnnexRepo(GitRepo, RepoInterface):
                     ':'.join(annex_cmd), annex_cmd,
                     path=self.path)(key_)
             except CommandError:
-                # git-annex will signal an unknown remote by exiting with
-                # return code 1.
+                # git-annex runs in batch mode, but will still signal some
+                # errors, e.g. an unknown remote, by exiting with a non-zero
+                # return code.
                 return False
             try:
                 return {

--- a/datalad/support/annexrepo.py
+++ b/datalad/support/annexrepo.py
@@ -2761,9 +2761,14 @@ class AnnexRepo(GitRepo, RepoInterface):
                 return False
         else:
             annex_cmd = ["checkpresentkey"] + ([remote] if remote else [])
-            out = self._batched.get(
-                ':'.join(annex_cmd), annex_cmd,
-                path=self.path)(key_)
+            try:
+                out = self._batched.get(
+                    ':'.join(annex_cmd), annex_cmd,
+                    path=self.path)(key_)
+            except CommandError:
+                # git-annex will signal an unknown remote by exiting with
+                # return code 1.
+                return False
             try:
                 return {
                     # happens on travis in direct/heavy-debug mode, that process

--- a/datalad/tests/test_cmd.py
+++ b/datalad/tests/test_cmd.py
@@ -170,7 +170,7 @@ exit(3)
     with pytest.raises(BatchedCommandError) as exception_info:
         bc("line two")
     assert exception_info.value.code == 3
-    assert exception_info.value.request == first_request
+    assert exception_info.value.last_processed_request == first_request
     assert bc.return_code == 3
 
     # Check for restart
@@ -194,7 +194,7 @@ print(a*b)
     with pytest.raises(BatchedCommandError) as exception_info:
         _ = bc(first_request)
     assert exception_info.value.code == 1
-    assert exception_info.value.request is None
+    assert exception_info.value.last_processed_request is None
     assert bc.return_code == 1
     assert bc.last_request is None
     bc.close(return_stderr=False)

--- a/datalad/tests/test_cmd.py
+++ b/datalad/tests/test_cmd.py
@@ -12,10 +12,14 @@ import sys
 import unittest.mock
 from subprocess import TimeoutExpired
 
+import pytest
+
 from datalad.cmd import (
     BatchedCommand,
     readline_rstripped,
 )
+from datalad.cmd import BatchedCommandError
+from datalad.runner.exception import CommandError
 from datalad.runner.tests.utils import py2cmd
 from datalad.tests.utils_pytest import (
     assert_equal,
@@ -148,8 +152,9 @@ def test_batched_restart():
 
 
 def test_command_fail_1():
-    # Expect that the return code of a failing command is caught,
-    # that None is returned as result.
+    # Expect that a failing command raises a CommandError in which the return
+    # code and the last successful request is caught, and that the command is
+    # restarted when called again
     bc = BatchedCommand(
         cmd=py2cmd(
             """
@@ -158,19 +163,29 @@ exit(3)
             """))
 
     # Send something to start the process
-    result = bc("line one")
-    assert_equal(bc.return_code, None)
-    assert_equal(result, "something")
-    result = bc("line two")
-    assert_equal(bc.return_code, 3)
-    assert_equal(result, None)
-    bc.close(return_stderr=False)
+    first_request = "line one"
+    result = bc(first_request)
+    assert bc.return_code == None
+    assert result == "something"
+    with pytest.raises(CommandError) as exception_info:
+        bc("line two")
+        assert exception_info.value.code == 3
+        print(exception_info.value.msg, file=sys.stderr)
+        assert exception_info.value.msg.endswith(first_request)
+
+    # Check for restart
+    #assert_equal(bc.return_code, None)
+    #result = bc(first_request)
+    #assert result == "something"
+    #with pytest.raises(CommandError) as exception_info:
+    #    bc.close(return_stderr=False)
+    #    assert exception_info.value.code == 3
 
 
 def test_command_fail_2():
-    # Expect that the return code of a failing command is caught,
-    # that None is returned as result and that the process is restarted,
-    # if the batched command is called again.
+    # Expect that a failing command raises a CommandError in which the return
+    # code and the last successful request is caught, and that the process is
+    # restarted if the batched command is called again
     bc = BatchedCommand(
         cmd=py2cmd(
             """
@@ -178,10 +193,26 @@ print(a*b)
             """))
 
     # Send something to start the process
-    result = bc("line one")
+    first_request = "line one"
+    result = bc(first_request)
     assert_not_equal(bc.return_code, 0)
     assert_is_none(result)
     result = bc("line two")
     assert_not_equal(bc.return_code, 0)
     assert_is_none(result)
+    bc.close(return_stderr=False)
+
+
+def test_command_fail_3():
+    # Expect that the return code of a failing command is caught,
+    # that None is returned as result.
+    bc = BatchedCommand(cmd=py2cmd("exit(4)"))
+
+    # Send something to start the process
+    result = bc("line one")
+    assert_equal(bc.return_code, None)
+    assert_equal(result, "something")
+    result = bc("line two")
+    assert_equal(bc.return_code, 3)
+    assert_equal(result, None)
     bc.close(return_stderr=False)

--- a/datalad/tests/test_cmd.py
+++ b/datalad/tests/test_cmd.py
@@ -165,27 +165,24 @@ exit(3)
     # Send something to start the process
     first_request = "line one"
     result = bc(first_request)
-    assert bc.return_code == None
+    assert bc.return_code is None
     assert result == "something"
-    with pytest.raises(CommandError) as exception_info:
+    with pytest.raises(BatchedCommandError) as exception_info:
         bc("line two")
-        assert exception_info.value.code == 3
-        print(exception_info.value.msg, file=sys.stderr)
-        assert exception_info.value.msg.endswith(first_request)
+    assert exception_info.value.code == 3
+    assert exception_info.value.request == first_request
+    assert bc.return_code == 3
 
     # Check for restart
-    #assert_equal(bc.return_code, None)
-    #result = bc(first_request)
-    #assert result == "something"
-    #with pytest.raises(CommandError) as exception_info:
-    #    bc.close(return_stderr=False)
-    #    assert exception_info.value.code == 3
+    result = bc(first_request)
+    assert result == "something"
+    bc.close(return_stderr=False)
 
 
 def test_command_fail_2():
-    # Expect that a failing command raises a CommandError in which the return
-    # code and the last successful request is caught, and that the process is
-    # restarted if the batched command is called again
+    # Expect that a failing command raises a BatchedCommandError in which the
+    # return code and the last successful request is caught. In this case the
+    # last successful request should be None.
     bc = BatchedCommand(
         cmd=py2cmd(
             """
@@ -194,25 +191,10 @@ print(a*b)
 
     # Send something to start the process
     first_request = "line one"
-    result = bc(first_request)
-    assert_not_equal(bc.return_code, 0)
-    assert_is_none(result)
-    result = bc("line two")
-    assert_not_equal(bc.return_code, 0)
-    assert_is_none(result)
-    bc.close(return_stderr=False)
-
-
-def test_command_fail_3():
-    # Expect that the return code of a failing command is caught,
-    # that None is returned as result.
-    bc = BatchedCommand(cmd=py2cmd("exit(4)"))
-
-    # Send something to start the process
-    result = bc("line one")
-    assert_equal(bc.return_code, None)
-    assert_equal(result, "something")
-    result = bc("line two")
-    assert_equal(bc.return_code, 3)
-    assert_equal(result, None)
+    with pytest.raises(BatchedCommandError) as exception_info:
+        _ = bc(first_request)
+    assert exception_info.value.code == 1
+    assert exception_info.value.request is None
+    assert bc.return_code == 1
+    assert bc.last_request is None
     bc.close(return_stderr=False)

